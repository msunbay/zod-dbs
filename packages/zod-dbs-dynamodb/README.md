# zod-dbs-dynamodb

DynamoDB provider for zod-dbs.

## Notes

- Requires `@aws-sdk/client-dynamodb` as a peer dependency.
- Enum types are not yet supported, but can be added to the generated schemas using hooks.
- DynamoDB is schemaless; results are heuristic (unioned types collapse to json/any).
- Key attributes forced non-nullable.
- No deep object/array field introspection inside M/L beyond coarse type classification.
- Requires AWS credentials or a local endpoint (set endpoint for DynamoDB local).

## Installation

```bash
npm install zod-dbs-dynamodb
```

> Note: This package has a peer dependency on `@aws-sdk/client-dynamodb`. If you don't already have it, you might need to install it (depending on your environment, package manager, etc):

```bash
npm install @aws-sdk/client-dynamodb
```

## CLI Usage

```bash
npx zod-dbs --provider dynamo --endpoint <endpoint> --sample-size <num>
```

### Provider Specific CLI Options

| Option          | Description                                                | Required |
| --------------- | ---------------------------------------------------------- | -------- |
| `--endpoint`    | Local endpoint                                             |          |
| `--sample-size` | Number of documents to sample per collection (default: 50) |          |

### Configuration File Example

```ts
import { ZodDbsCliConfig } from 'zod-dbs-cli';

// Import needed to load the provider specific configuration types.
import 'zod-dbs-dynamodb';

const config: ZodDbsCliConfig = {
  provider: 'dynamo',
  endpoint: '<endpoint>', // optional for local DynamoDB
  sampleSize: 100, // optional default is 50
};

export default config;
```

## Programmatic Usage

```ts
import { generateZodSchemas } from 'zod-dbs';
import { createProvider } from 'zod-dbs-dynamodb';

await generateZodSchemas({
  provider: createProvider(),
  config: {
    endpoint: '<endpoint>', // optional for local DynamoDB
    sampleSize: 50, // optional default is 50
  },
});
```
