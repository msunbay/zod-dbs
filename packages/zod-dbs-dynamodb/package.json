{
  "name": "zod-dbs-dynamodb",
  "version": "0.0.1",
  "description": "DynamoDB provider for zod-dbs",
  "main": "./dist/index.js",
  "types": "./dist/index.d.ts",
  "type": "module",
  "scripts": {
    "build": "rm -rf ./dist && tsc --project tsconfig.build.json",
    "test:integration": "vitest --project integration run",
    "test:types": "tsc --noEmit"
  },
  "author": "msolvaag",
  "license": "ISC",
  "repository": {
    "type": "git",
    "url": "git+https://github.com/msolvaag/zod-dbs.git",
  "directory": "packages/zod-dbs-dynamodb"
  },
  "publishConfig": {
    "access": "public"
  },
  "files": [
    "dist"
  ],
  "dependencies": {
    "zod-dbs-core": "workspace:*"
  },
  "peerDependencies": {
    "@aws-sdk/client-dynamodb": ">=3"
  },
  "devDependencies": {
    "@aws-sdk/client-dynamodb": "^3.879.0"
  }
}
