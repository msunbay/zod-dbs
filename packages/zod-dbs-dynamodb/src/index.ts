export { DynamoDbProvider } from './DynamoDbProvider.js';
export { createProvider } from './provider.js';
export { createClient } from './client.js';
export type * from './types.js';
