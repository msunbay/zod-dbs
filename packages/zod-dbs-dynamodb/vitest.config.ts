import { defineConfig } from 'vitest/config';

export default defineConfig({
  test: {
    reporters: 'dot',
    projects: [
      {
        test: {
          name: 'unit',
          environment: 'node',
          include: ['test/unit/**/*.test.ts'],
          globals: true,
        },
      },
    ],
  },
});
