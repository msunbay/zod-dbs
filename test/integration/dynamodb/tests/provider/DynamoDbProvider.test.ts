import { DynamoDbProvider } from 'zod-dbs-dynamodb';

import { getProviderConfig } from '../../../utils/context.js';

it('returns inferred schema column information', async () => {
  const provider = new DynamoDbProvider();
  const columns = await provider.fetchSchemaInfo(getProviderConfig());

  expect(columns).toBeDefined();
  expect(columns.length).toBeGreaterThan(0);

  const usersCols = columns.filter((c) => c.tableName === 'Users');
  expect(usersCols.length).toBeGreaterThan(0);

  const idCol = usersCols.find((c) => c.name === 'id');
  expect(idCol).toBeDefined();
  expect(idCol!.isNullable).toBe(false);

  // Clean undefined for snapshot stability
  usersCols.forEach((c) => {
    (Object.keys(c) as (keyof typeof c)[]).forEach((k) => {
      if (c[k] === undefined) delete c[k];
    });
  });

  expect(usersCols).toMatchSnapshot('usersColumns');
});

it('returns table models', async () => {
  const provider = new DynamoDbProvider();
  const info = await provider.getSchemaInformation(getProviderConfig());

  expect(info.tables.length).toBeGreaterThan(0);
  const users = info.tables.find((t) => t.name === 'Users');
  expect(users).toBeDefined();
  expect(users!.columns.length).toBeGreaterThan(0);
  expect(users).toMatchSnapshot('usersTable');
});
