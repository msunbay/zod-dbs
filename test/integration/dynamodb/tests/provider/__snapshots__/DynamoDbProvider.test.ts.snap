// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`returns inferred schema column information > usersColumns 1`] = `
[
  {
    "dataType": "string",
    "isArray": false,
    "isEnum": false,
    "isNullable": false,
    "isSerial": false,
    "name": "id",
    "tableName": "Users",
    "tableType": "table",
  },
  {
    "dataType": "string",
    "isArray": false,
    "isEnum": false,
    "isNullable": false,
    "isSerial": false,
    "name": "sort",
    "tableName": "Users",
    "tableType": "table",
  },
  {
    "dataType": "string",
    "isArray": false,
    "isEnum": false,
    "isNullable": false,
    "isSerial": false,
    "name": "email",
    "tableName": "Users",
    "tableType": "table",
  },
  {
    "dataType": "object",
    "isArray": false,
    "isEnum": false,
    "isNullable": false,
    "isSerial": false,
    "name": "profile",
    "tableName": "Users",
    "tableType": "table",
  },
  {
    "dataType": "array",
    "isArray": false,
    "isEnum": false,
    "isNullable": false,
    "isSerial": false,
    "name": "tags",
    "tableName": "Users",
    "tableType": "table",
  },
]
`;

exports[`returns table models > usersTable 1`] = `
{
  "columns": [
    {
      "dataType": "string",
      "defaultValue": undefined,
      "description": undefined,
      "enumValues": undefined,
      "isArray": false,
      "isEnum": false,
      "isNullable": false,
      "isReadOptional": false,
      "isSerial": false,
      "isWritable": true,
      "isWriteOptional": false,
      "maxLen": undefined,
      "minLen": undefined,
      "name": "id",
      "schemaName": undefined,
      "tableName": "Users",
      "tableType": "table",
      "type": "string",
    },
    {
      "dataType": "string",
      "defaultValue": undefined,
      "description": undefined,
      "enumValues": undefined,
      "isArray": false,
      "isEnum": false,
      "isNullable": false,
      "isReadOptional": false,
      "isSerial": false,
      "isWritable": true,
      "isWriteOptional": false,
      "maxLen": undefined,
      "minLen": undefined,
      "name": "sort",
      "schemaName": undefined,
      "tableName": "Users",
      "tableType": "table",
      "type": "string",
    },
    {
      "dataType": "string",
      "defaultValue": undefined,
      "description": undefined,
      "enumValues": undefined,
      "isArray": false,
      "isEnum": false,
      "isNullable": false,
      "isReadOptional": false,
      "isSerial": false,
      "isWritable": true,
      "isWriteOptional": false,
      "maxLen": undefined,
      "minLen": undefined,
      "name": "email",
      "schemaName": undefined,
      "tableName": "Users",
      "tableType": "table",
      "type": "string",
    },
    {
      "dataType": "object",
      "defaultValue": undefined,
      "description": undefined,
      "enumValues": undefined,
      "isArray": false,
      "isEnum": false,
      "isNullable": false,
      "isReadOptional": false,
      "isSerial": false,
      "isWritable": true,
      "isWriteOptional": false,
      "maxLen": undefined,
      "minLen": undefined,
      "name": "profile",
      "schemaName": undefined,
      "tableName": "Users",
      "tableType": "table",
      "type": "object",
    },
    {
      "dataType": "array",
      "defaultValue": undefined,
      "description": undefined,
      "enumValues": undefined,
      "isArray": false,
      "isEnum": false,
      "isNullable": false,
      "isReadOptional": false,
      "isSerial": false,
      "isWritable": true,
      "isWriteOptional": false,
      "maxLen": undefined,
      "minLen": undefined,
      "name": "tags",
      "schemaName": undefined,
      "tableName": "Users",
      "tableType": "table",
      "type": "any",
    },
  ],
  "name": "Users",
  "schemaName": undefined,
  "type": "table",
}
`;
